components/brave_extension/extension/brave_extension/background/greaselion.ts

```diff
@@ -129,19 +158,12 @@ const handleRegisterOnCompletedWebRequest = (registrationKey: string, mediaType:
   // Mark this tab as registered
   onCompletedWebRequestRegistrations.set(registrationKey, { registered: true })
 
+  // Create and install the listener
+  onCompletedWebRequestBinding =
```

Comment:
Will each new call to `handleRegisterOnCompletedWebRequest ` overwrite the `onCompletedWebRequestBinding` variable? If so, then does that mean that overwritten `onCompletedWebRequestBinding` listeners will not be removed when the port is disconnected?

Instead of overwriting these module-scoped variables, we could maybe store the listeners in the registration maps (e.g. `onCompletedWebRequestRegistrations`).

Another option: since all of these listeners are scoped to the port connection, we could also store all of the per-connection data in the `portsByTabIdSenderId` map. Sorta like:

```ts
interface ConnectionState {
  port: chrome.runtime.Port
  onUpdatedTabListener: (tabId: number, changeInfo: chrome.tabs.TabChangeInfo, tab: chrome.tabs.Tab) => void
  onSendHeadersWebRequestListener: (details: chrome.webRequest.WebRequestHeadersDetails) => void
  // etc...
}

const portsByTabIdSenderId = new Map<string, ConnectionState>()
```

Reply:
Ah yeah, that will be a problem. Strangely enough, it seemed to be doing the right thing when I was testing. Maybe I wasn't exercising it properly though. Will go with the `ConnectionState` approach you mentioned above, that seems like a great idea - thank you!

Reply:
Fixed, thanks for the suggestion!